adding generate params command

polatengin opened the pull request:
Fixes #512 

## Contributing a feature

* [x] I have opened a new issue for the proposal, or commented on an existing one, and ensured that the Bicep maintainers are good with the design of the feature being implemented
* [x] I have included "Fixes #{issue_number}" in the PR description, so GitHub can link to the issue and close it when the PR is merged

/cc: @majastrz , @ucheNkadiCode

alex-frankel commented:
Thanks for the contribution @polatengin!

Can you clarify the intended usage of the feature? I looked in the command bar in VS code and tried right-clicking on the file name, but didn't see anything in either location. Is this implemented in the CLI?

ucheNkadiCode commented:
@shenglol  and @anthony-c-martin when either of you have time.

From my understanding, the updates that were made were to make this command available in the Bicep CLI as well as the Command Pallete.  Thank you!

Questions for @polatengin, 

1. I also believe this overwrites a parameters file of the same name each time (e.g `main.parameters.json`) it is called correct Engin? Or does it update the file to append params that are not found and keeps all of the existing params? 
2. Does this also implement telemetry so we can tell how often customers use this feature? Thank you!

shenglol commented on src/vscode-bicep/src/extension.ts:
```diff
@@ -12,7 +12,7 @@ import {
   BuildCommand,
   CommandManager,
   DeployCommand,
-  ForceModulesRestoreCommand,
```
Need to restore `ForceModulesRestoreCommand`.

polatengin replied:
just put it back, thanks 👍

bhsubra commented:
@polatengin, as discussed offline, on the vscode side, we send out telemetry event when a command is invoked. We also capture errors. 

This is with your changes:
![image](https://user-images.githubusercontent.com/30270536/169177008-83261690-f187-47e9-9698-15631e587ede.png)

@ucheNkadiCode, do we want to capture anything else on vscode side? Regarding cli, we don't have infrastructure in place to capture telemetry data. Should we create a separate issue to handle that?

majastrz commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,60 @@
+
+            this.EmitMetadata(jsonWriter, emitter);
+
+            this.EmitParametersIfPresent(jsonWriter, emitter);
```
>EmitParametersIfPresent [](http://example.com/codeflow?start=17&length=23)

Are you sure this is the right method to call here? This is the method from the template writer that writes out the parameter schema including `type` properties which don't belong in the parameters file, which should only have parameter values.

shenglol replied:
+1

polatengin replied:
you're right, and according to the official 'parameters file' documentation, the json file shouldn't include 'type' fields.

I sent an update to the [ParametersFileTemplateWriter.cs](https://github.com/Azure/bicep/pull/6601/files/c4ef9ec5347315890d6c501f7adb5c793fcd5c0c#diff-30a900071270f3f38861de1729f2ccb8d1776b6467b266f561c7b255a50eaba3) and changed the way to emit parameter fields.

That change should resolve your comment, thanks 👍

majastrz commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+                if (!string.IsNullOrEmpty(template))
+                {
+                    JToken jtoken = template.FromJson<JToken>();
+                    if (TemplateHelpers.TryGetTemplateGeneratorObject(jtoken, out DeploymentTemplateGeneratorMetadata generator))
```
>TryGetTemplateGeneratorObject [](http://example.com/codeflow?start=40&length=29)

Do param files actually allow open-ended `metadata` like templates do? I don't see it documented here [https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/parameter-files](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/parameter-files), but I could be missing something?

polatengin replied:
I tried to deploy a generated arm template with `metadata` structure in the parameters.json file, it went successfully and deployed.

I decided to keep the `metadata` structure to have some information in the parameters file too.

if you see it's not a good fit, I can remove it.

ucheNkadiCode replied:
@polatengin  I would like for "metadata" to be removed. It doesn't fit into our best practices

majastrz commented:
You should also consider adding integration tests that generate param files and compare them against checked in param files.

shenglol commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,60 @@
+        public new void Write(JsonTextWriter writer)
+        {
+            // Template is used for calcualting template hash, template jtoken is used for writing to file.
+            var templateJToken = GenerateTemplate();
```
I wonder if we need to generate a template here. Since we have access to the semantic model, we should be able to generate parameters using the parameter type information.

majastrz replied:
Yeah, I agree.

polatengin commented:
> Thanks for the contribution @polatengin!
> 
> Can you clarify the intended usage of the feature? I looked in the command bar in VS code and tried right-clicking on the file name, but didn't see anything in either location. Is this implemented in the CLI?

somehow I missed this question; this PR is to add 'generate commands' feature to both CLI and VSCode extension

polatengin commented on src/Bicep.Core/Emit/TemplateWriter.cs:
```diff
@@ -144,7 +144,7 @@ public void Write(JsonTextWriter writer)
         }
 
-        private void EmitParametersIfPresent(JsonTextWriter jsonWriter, ExpressionEmitter emitter)
+        protected void EmitParametersIfPresent(JsonTextWriter jsonWriter, ExpressionEmitter emitter)
```
just created a new commit and revert back the access modifier, thanks 👍

polatengin commented:
> You should also consider adding integration tests that generate param files and compare them against checked in param files.

yes, currently I'm working on to add some tests

ucheNkadiCode commented:
> @polatengin, as discussed offline, on the vscode side, we send out telemetry event when a command is invoked. We also capture errors.
> 
> This is with your changes: ![image](https://user-images.githubusercontent.com/30270536/169177008-83261690-f187-47e9-9698-15631e587ede.png)
> 
> @ucheNkadiCode, do we want to capture anything else on vscode side? Regarding cli, we don't have infrastructure in place to capture telemetry data. Should we create a separate issue to handle that?

I think we are good when it comes to telemetry here

ucheNkadiCode commented:
> > > 1. I also believe this overwrites a parameters file of the same name each time (e.g `main.parameters.json`) it is called correct Engin? Or does it update the file to append params that are not found and keeps all of the existing params?
> > 
> > 
> > Yes, it overwrites the parameters file, if there is already one. if this is not the expected behavior, I can change it.
> 
> I think it would be really cool if we could generate the param file non-destructively. Basically, we could read in the existing param file (if any), capture the param values, and put them in the generated one. This way the user wouldn't lose any data accidentally.

Just making sure @polatengin doesn't miss this. I think through the suggestions of our team, it would be great if we simply append to the already existing parameters.json file. But we'll make sure to chat about it in our discussion today

polatengin commented:
> You should also consider adding integration tests that generate param files and compare them against checked in param files.

I just added a few integration tests, during adding those tests, I noticed an issue and fixed it, too 👍

anthony-c-martin commented on src/Bicep.Cli/Commands/GenerateParametersFileCommand.cs:
```diff
@@ -0,0 +1,77 @@
+
+namespace Bicep.Cli.Commands
+{
+    public class GenerateParametersFileCommand : ICommand
```
Please make sure you add documentation for this new command to the CLI help text:

https://github.com/Azure/bicep/blob/8004f3a1ae0179018d273d8df73f9490e23e4db3/src/Bicep.Cli/Commands/RootCommand.cs#L66-L108

ucheNkadiCode commented on src/Bicep.Cli/Commands/RootCommand.cs:
```diff
@@ -101,6 +101,26 @@ private void PrintHelp()
+    Options:
+      --no-restore      Generates the parameters file without restoring external modules.
+      --outdir <dir>    Saves the output at the specified directory.
+      --outfile <file>  Saves the output as the specified file path.
```
For outfile, should this be "Saves the output as the specified file name" -? Would that be more accurate @polatengin

polatengin replied:
I think that makes more sense. but, I was following the same wording with the `build` https://github.com/Azure/bicep/blob/53254fb484b24b5a3746ed01e3cbc5acfcf3e256/src/Bicep.Cli/Commands/RootCommand.cs#L75 and `decompile` https://github.com/Azure/bicep/blob/53254fb484b24b5a3746ed01e3cbc5acfcf3e256/src/Bicep.Cli/Commands/RootCommand.cs#L94 help texts. they have `Saves the output as the specified file path.` for the `--outfile` option.

what you think? should I leave it as-is, only update it for `generate-params` command, or update every instance?

StephenWeatherford replied:
Since they can specify a full path (right?), "file path" is more technically correct.  If so, I vote for leave as is.

My assumption is that you can't specify both --outdir and --outfile, correct?

polatengin replied:
yes, they don't work together, one or the other

ucheNkadiCode replied:
Okay Leave as is

anthony-c-martin commented on src/Bicep.LangServer/LangServerConstants.cs:
```diff
@@ -6,6 +6,7 @@ namespace Bicep.LanguageServer
     public static class LangServerConstants
     {
         public const string BuildCommand = "build";
+        public const string GenerateParamsCommand = "generateParams";
```
Please add a test for this new command - you could probably base this off https://github.com/Azure/bicep/blob/7d2b7818e0bd0432560f46713a8584dc21fa12bc/src/Bicep.LangServer.IntegrationTests/BuildCommandTests.cs#L22-L23

polatengin replied:
done! added a test to check if generate-params command works as expected 👍

bhsubra commented:
@polatengin, az deployment and vscode deploy command supports parameters file with both .json/.jsonc extension. With the changes in this PR, looks like if there is a parameters file with .jsonc extension, then it's ignored and we go ahead and create another file with .json extension. Is that intentional? 
If we plan to support .jsonc file, the logic you have for merging might not work as it removes comments while serialization. We could use https://github.com/Azure/bicep/blob/main/src/Bicep.LangServer/Configuration/JsonEditor.cs to update json instead, which has logic to preserve comments.

cc: @ucheNkadiCode

StephenWeatherford commented on src/Bicep.Cli/Commands/RootCommand.cs:
```diff
@@ -101,6 +101,26 @@ private void PrintHelp()
       bicep decompile file.json --force
       bicep decompile file.json --outfile file.bicep
+
+  {exeName} generate-params [options] <file>
```
Documentation should indicate what happens if the file already exists.

polatengin replied:
created a commit to update the help text

StephenWeatherford commented on src/Bicep.Cli/Services/ParametersFileWriter.cs:
```diff
@@ -0,0 +1,58 @@
+            var fileStream = CreateFileStream(outputPath);
+            using (fileStream)
+            {
+                return new TemplateEmitter(compilation.GetEntrypointSemanticModel(), invocationContext.EmitterSettings).EmitParametersFile(fileStream);
```
Why not call ToStream method?

StephenWeatherford commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+                throw new InvalidOperationException($"generated template doesn't contain a generator object at the path {GeneratorMetadataPath}");
+            }
+
+            if (!string.IsNullOrWhiteSpace(existingContent))
```
Does the user get a reasonable error experience if the parse fails?

polatengin replied:
current infrastructure to handle errors will catch the exception and handle it properly. I based off this code from the `TemplateWriter` class
 https://github.com/polatengin/bicep/blob/dbe1bff5efa6a38419a9330c7e54cae85f84dfc1/src/Bicep.Core/Emit/TemplateWriter.cs#L104

StephenWeatherford commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+
+                        if (existingParameterValue is not null)
+                        {
+                            if (templateJToken.SelectToken($"parameters.{existingParameterName}") is JObject existingParameterObject)
```
Probably don't care, but might be issues here with oddly-named parameters.

polatengin replied:
what you mean? like non-ANSI characters?

StephenWeatherford replied:
Hyphens, slashes etc.  Technically I think they're allowed by ARM, but I don't think Bicep can access them, so probably okay.

StephenWeatherford replied:
I'm going to close for now.

polatengin replied:
thanks for the information, I'm clicking "resolve conversation" button for now, if you think we should address oddly-named parameters, in the future, we can create an issue and a follow-up PR 👍

StephenWeatherford commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+                        {
+                            if (templateJToken.SelectToken($"parameters.{existingParameterName}") is JObject existingParameterObject)
+                            {
+                                existingParameterObject.Merge(existingParameterValue, mergeSettings);
```
Can this fail?  Can we be left in a state where part of the new file is written out, then fails, leaving the new file invalid?

polatengin replied:
it shouldn't fail. but added a try-catch block, just in case, if the existing params content is not valid json

StephenWeatherford commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+
+                        if (existingParameterValue is not null)
+                        {
+                            if (templateJToken.SelectToken($"parameters.{existingParameterName}") is JObject existingParameterObject)
```
Parameter names (technically *all* property names, but we can probably ignore that for other cases) are case-insensitive, so you need to match case-insensitively.

StephenWeatherford commented on src/Bicep.Cli/Commands/RootCommand.cs:
```diff
@@ -101,6 +101,26 @@ private void PrintHelp()
+      bicep generate-params file.bicep --stdout
+      bicep generate-params file.bicep --outdir dir1
+      bicep generate-params file.bicep --outfile file.parameters.json
+
```
@ucheNkadiCode Would it be better to have a specific option to do merging of an existing params file?  E.g. bicep generate-params file.bicep --update

polatengin replied:
if I remember correctly, the consensus was that users are already using source control mechanism, if somehow the end result is not what they want, they can use git techniques to revert changes. to keep simplicity of the code and behavior, it should always behave the same way.

ucheNkadiCode replied:
@shenglol, @anthony-c-martin , @majastrz do you have an opinion here? My thoughts are if the file exists and they type this command into cli, it would make sense to overwrite it here

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+            string compiledFilePath = PathHelper.ResolveParametersFileOutputPath(bicepFilePath);
+            string compiledFile = Path.GetFileName(compiledFilePath);
+
+            // If the template exists and contains bicep generator metadata, we can go ahead and replace the file.
```
TODO

StephenWeatherford replied:
Also, it shows a validation error in the vscode editor:
![image](https://user-images.githubusercontent.com/6913354/171500780-6eed8a11-d9cc-4380-8b1a-339be78b57ae.png)

polatengin replied:
to match the generated json and parameters file, I add the metadata into the parameters file too. per my tests, azure cli successfully created deployment, even if the parameters file has metadata field in it.

I couldn't understand what you mean by `if you start with your own params file you can't use this command`, it must parse the existing parameters file, and use the `values`, `contentVersion`, etc. off of it.

you're right about the vscode extension, though.

@ucheNkadiCode , what you think?

StephenWeatherford replied:
@ucheNkadiCode Let's discuss in person at our sync.

StephenWeatherford replied:
Remove metadata.

polatengin replied:
done! metadata field is removed. tests are updated to reflect changes. $schema field used to validate the generated json file, instead of metadata field.

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+            var emitter = new TemplateEmitter(compilation.GetEntrypointSemanticModel(), emitterSettings);
+            EmitResult result = emitter.EmitParametersFile(fileStream);
+
+            return "Generate Params succeeded. Created file " + compiledFile;
```
If the file already exists, it should say "Updated file"...

polatengin replied:
changed the message to `processed file`

StephenWeatherford replied:
I guess good enough for now.

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        private string GenerateCompiledParametersFileAndReturnOutputMessage(string bicepFilePath, DocumentUri documentUri)
+        {
```
@ucheNkadiCode Given this bicep file:
```bicep
param nodefault string
@description('hello')
param hasdefault string = 'hello'
```
The params generated is this:
```json
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "metadata": {
    "_generator": {
      "name": "bicep",
      "version": "0.6.68.62354"
    }
  },
  "parameters": {
    "nodefault": {},
    "hasdefault": {
      "value": "hello"
    }
  }
}
```
This writes out the default values from the bicep file into the params file.  I don't think that's the right thing to do - I think by parameters with default values should be left out of the param file so that they actually pick up the default in the bicep file if it changes.

The ARM extension actually asks whether to write out all params or only those without default values:
![image](https://user-images.githubusercontent.com/6913354/171502170-18c5376e-8754-47b7-9dab-88b0128fba35.png)
Either way, it does *not* copy over the default values, e.g.:
Template:
```json
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "nodefault": {
            "type": "string",
            "metadata": {
                "description": "description"
            }
        },
        "hasdefault": {
            "type": "string",
            "defaultValue": "my default value"
        }
    },
    "resources": []
}
```
Generated params file with "all parameters" option:
```json
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "nodefault": {
            "value": "" // TODO: Fill in parameter value
        },
        "hasdefault": {
            "value": "my default value"
        }
    }
}
```
Generated params file with "only required parameters" option:
```json
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "nodefault": {
            "value": "" // TODO: Fill in parameter value
        }
    }
}
```

StephenWeatherford replied:
Just as a side note, the ARM extension takes a different approach to updating params files...  It uses a lightbulb:
![image](https://user-images.githubusercontent.com/6913354/171503618-52d6193b-a1ba-486b-b7dc-dc49059e27ee.png)

StephenWeatherford replied:
@ucheNkadiCode Let's discuss at sync.

StephenWeatherford replied:
TODO: Discuss on Monday

StephenWeatherford replied:
Decision for now: Just skip those with expressions as default values for now.

polatengin replied:
as we discussed, if parameter has a default value defined in the bicep file, generate-params command ignore it when building the parameters.json file

StephenWeatherford commented:
Reviewed a lot of it but not all of it yet, but that's enough comments for now :-).

StephenWeatherford commented on src/Bicep.Core/Emit/TemplateEmitter.cs:
```diff
@@ -27,6 +27,35 @@ public TemplateEmitter(SemanticModel model, EmitterSettings settings)
+        {
+            // read content of the stream into a string
+            using var reader = new StreamReader(stream);
+            var content = reader.ReadToEndAsync().GetAwaiter().GetResult();
```
GetResult() will block the thread, seems like you should use an async function with await.

StephenWeatherford replied:
If it's difficult to implement, please ask @majastrz  or @anthony-c-martin.

StephenWeatherford replied:
Use the sync version instead:

var content = reader.ReadToEnd();

polatengin replied:
done! thanks...

polatengin commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            // Template is used for calcualting template hash, template jtoken is used for writing to file.
+            var templateJToken = GenerateTemplate();
+            if (templateJToken.SelectToken(GeneratorMetadataPath) is not JObject generatorObject)
```
leftover from my initial experiments, removed the variable

polatengin commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+            var templateJToken = GenerateTemplate();
+            if (templateJToken.SelectToken(GeneratorMetadataPath) is not JObject generatorObject)
+            {
+                throw new InvalidOperationException($"generated template doesn't contain a generator object at the path {GeneratorMetadataPath}");
```
good suggestion, words added..

StephenWeatherford replied:
Has this been pushed?  Don't see it

polatengin replied:
we had it, but since we decided to get rid of generator metadata, the code is removed. no need to update the text anymore. resolving comment, thanks!..

polatengin commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+
+            emitter.EmitProperty("$schema", "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#");
+
+            emitter.EmitProperty("contentVersion", "1.0.0.0");
```
good suggestion, changed the code a bit to handle this scenario. also added fail-safe to see if the content of the existing params file is valid or invalid.

added couple of tests to see if this behaviors are working as expected

polatengin commented on src/vscode-bicep/src/commands/generateParams.ts:
```diff
@@ -0,0 +1,45 @@
+      );
+      this.outputChannelManager.appendToOutputChannel(generateParamsOutput);
+    } catch (err) {
+      this.client.error(
```
I based the `generateParams.ts` off from the `build.ts` file, and it's the way to re-throw error in the catch block;

https://github.com/polatengin/bicep/blob/enpolat/512-generate-params-command/src/vscode-bicep/src/commands/build.ts#L37

StephenWeatherford replied:
wait for feedback

StephenWeatherford replied:
The other code is questionable.  Please fix.  Change client.error to throw new Error()

polatengin replied:
done, as we speak, thanks 👍

polatengin commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+            // If not, we'll fail the generate params.
+            if (File.Exists(compiledFilePath) && !TemplateContainsBicepGeneratorMetadata(File.ReadAllText(compiledFilePath)))
+            {
+                return "Generate Params failed. The file \"" + compiledFile + "\" already exists and was not generated by Bicep. If overwriting the file is intended, delete it manually and retry the Generate Params command.";
```
done! thanks..

StephenWeatherford replied:
Is this pushed?

StephenWeatherford replied:
P2

polatengin replied:
done, thanks!.. 👍

ucheNkadiCode commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,103 @@
+
+            if (!string.IsNullOrWhiteSpace(existingContent))
+            {
+                var fileJToken = JToken.Parse(existingContent);
```
I think we should skip them for now to be consistent with the Deploy command. We can update that behavior with user feedback. For now @polatengin, would you mind skipping params with the `@secure` decorator?

StephenWeatherford replied:
Agreed:  Please skip secure parameters for now.

polatengin replied:
thanks for the information, resolving this conversation for now, when we have user feedback that says otherwise, we can create a follow-up PR to address this issue 👍

StephenWeatherford commented on src/Bicep.Core/Emit/ParametersFileTemplateWriter.cs:
```diff
@@ -0,0 +1,117 @@
+                }
+                catch
+                {
+                    // content of the existing file is not valid json, ignore merging it
```
Thanks, but I think it would be better to throw an error.  It will be a surprise for the user to have their file overwritten if there are (unbeknownst them to) errors in it.

StephenWeatherford replied:
P2

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+            return Task.FromResult(output);
+        }
+
+        private string GenerateCompiledParametersFileAndReturnOutputMessage(string bicepFilePath, DocumentUri documentUri)
```
@ucheNkadiCode Let's discuss at sync.

ucheNkadiCode replied:
Remove metadata from output

StephenWeatherford replied:
Decision: Put value in:

 "parameters": {
    "a": {
      "value": ""
    },
    "b": {
      "value": 0   << empty array or object as appropriate
    },

miqm commented on src/Bicep.Core/Emit/TemplateEmitter.cs:
```diff
@@ -27,6 +27,35 @@ public TemplateEmitter(SemanticModel model, EmitterSettings settings)
+        public EmitResult EmitParametersFile(Stream stream) => EmitOrFail(() =>
+        {
+            // read content of the stream into a string
+            using var reader = new StreamReader(stream);
```
why do we need to read the stream?

StephenWeatherford commented on src/Bicep.Cli.IntegrationTests/GenerateParamsCommandTests.cs:
```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Microsoft Corporation.
```
Bug

StephenWeatherford replied:
P1

polatengin replied:
found the buggy code and fixed it 🥳🥳🥳

StephenWeatherford commented on src/Bicep.Cli.IntegrationTests/GenerateParamsCommandTests.cs:
```diff
@@ -0,0 +1,347 @@
+// Copyright (c) Microsoft Corporation.
```
Please (from the vscode only) open the updated/generated params file.

polatengin replied:
done! thanks..

StephenWeatherford commented on src/Bicep.Cli.IntegrationTests/GenerateParamsCommandTests.cs:
```diff
@@ -0,0 +1,347 @@
+// Copyright (c) Microsoft Corporation.
```
Skip putting parameters into the params file that have a default value that is an expression

StephenWeatherford commented:
Getting close, thanks!  I'll be out Tue-Fri, anthony-c-martin or ucheNkadiCode or bhsubra can help after that if needed.

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,146 @@
+                    JToken jtoken = template.FromJson<JToken>();
+                    var schema = jtoken.SelectToken("$schema")?.ToObject<JObject>();
+
+                    if (schema != null)
```
Many JSON files contain schemas, you need to check that the schema contains the string "deploymentParameters" (non-case-sensitive).

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,146 @@
+            // If not, we'll fail the generate params.
+            if (File.Exists(compiledFilePath) && !TemplateContainsSchemaField(File.ReadAllText(compiledFilePath)))
+            {
+                return "Generating parameters file failed. The file \"" + compiledFile + "\" already exists and was not generated by Bicep. If overwriting the file is intended, delete it manually and retry the Generate Params command.";
```
```suggestion
                return "Generating parameters file failed. The file \"" + compiledFile + "\" already exists but does not contain the schema for a parameters file. If overwriting the file is intended, delete it manually and retry the Generate Params command.";
```

polatengin replied:
updated the text message, thanks 👍

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,146 @@
+            return "Generating parameters file succeeded. Processed file " + compiledFile;
+        }
+
+        // Returns true if the template contains $schema field, false otherwise
```
```suggestion
        // Returns true if the template contains the parameters file schema, false otherwise
```

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        // Returns true if the template contains $schema field, false otherwise
+        public bool TemplateContainsSchemaField(string template)
```
suggested: "TemplateIsParametersFile"

polatengin replied:
thanks for the suggestion, fair.. made the change..

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -119,20 +119,19 @@ private string GenerateCompiledParametersFileAndReturnOutputMessage(string bicep
                 {
                     JToken jtoken = template.FromJson<JToken>();
-                    if (TemplateHelpers.TryGetTemplateGeneratorObject(jtoken, out DeploymentTemplateGeneratorMetadata generator))
+                    var schema = jtoken.SelectToken("$schema")?.ToObject<JObject>();
```
```suggestion
                    var schema = jtoken.SelectToken("$schema")?.ToString();
```

StephenWeatherford commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -119,20 +119,19 @@ private string GenerateCompiledParametersFileAndReturnOutputMessage(string bicep
-                    if (TemplateHelpers.TryGetTemplateGeneratorObject(jtoken, out DeploymentTemplateGeneratorMetadata generator))
+                    var schema = jtoken.SelectToken("$schema")?.ToObject<JObject>();
+
+                    if (schema != null)
```
```suggestion
                    if (schema != null && schema.ContainsInsensitively("deploymentParameters.json"))
```

StephenWeatherford commented on src/vscode-bicep/src/commands/generateParams.ts:
```diff
@@ -35,11 +35,7 @@ export class GenerateParamsCommand implements Command {
-        parseError(err).message,
-        true
-      );
+      new Error(`Generating parameters failed: ${parseError(err).message}`);
```
```suggestion
      throw new Error(
        `Generating parameters failed: ${parseError(err).message}`
      );
```
Forgot the throw

StephenWeatherford commented on src/vscode-bicep/src/commands/generateParams.ts:
```diff
@@ -34,6 +35,15 @@ export class GenerateParamsCommand implements Command {
+
+      const filePath = path.parse(documentUri.fsPath);
+
+      const openPath = vscode.Uri.parse(
```
The best way to handle this is to have the generateParams command return the path that it created/updated. Then there's one source of truth.

If you don't choose to go that route, you need to be careful about parsing URIs in vscode/typescript.

Use instead this (approproximately):
```
const openPath = Uri.file(path.join(filePath.dir, `${filePath.name}.parameters.json`));
```

polatengin replied:
I'd like to go with the latter approach, since I don't have enough knowledge to change the return value of the `workspace/executeCommand` function. I may not even understand how this code works, yet.

Created commit to update the approach with the suggestion, thanks..

StephenWeatherford commented on src/vscode-bicep/src/commands/generateParams.ts:
```diff
@@ -34,6 +35,15 @@ export class GenerateParamsCommand implements Command {
+      const filePath = path.parse(documentUri.fsPath);
+
+      const openPath = vscode.Uri.parse(
+        `file://${filePath.dir}/${filePath.name}.parameters.json`
```
better: 
var doc = await vscode.workspace.openTextDocument(openPath);
await vscode.window.showTextDocument(doc);

polatengin replied:
that totally makes sense, thanks for the suggestion. created commit to update the code

anthony-c-martin commented on src/Bicep.LangServer/Handlers/BicepGenerateParamsCommandHandler.cs:
```diff
@@ -0,0 +1,145 @@
+            }
+
+            var emitter = new TemplateEmitter(compilation.GetEntrypointSemanticModel(), emitterSettings);
+            using var fileStream = new FileStream(compiledFilePath, FileMode.OpenOrCreate, FileAccess.ReadWrite);
```
I don't think `FileMode.OpenOrCreate` is what you want here - it means you'll potentially only overwrite part of the file.

anthony-c-martin reviewed (approved):
Looks good, thanks for contributing!
